language/move-vm/types/src/values/values_impl.rs

```diff
@@ -2178,57 +2184,136 @@ impl Struct {
-                    status: Rc::new(RefCell::new(GlobalDataStatus::Clean)),
-                    fields: r,
-                })
+impl GlobalValueImpl {
+    fn cached(val: ValueImpl, status: GlobalDataStatus) -> PartialVMResult<Self> {
+        match val {
+            ValueImpl::Container(Container::StructR(fields)) => {
+                let status = Rc::new(RefCell::new(status));
+                Ok(Self::Cached { fields, status })
             }
-            v => Err(
+            _ => Err(
                 PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
-                    .with_message(format!("cannot create global ref from {:?}", v)),
+                    .with_message("failed to publish cached: not a resource".to_string()),
             ),
         }
     }
 
-    pub fn borrow_global(&self) -> PartialVMResult<Value> {
-        Ok(Value(ValueImpl::ContainerRef(ContainerRef::Global {
-            status: Rc::clone(&self.status),
-            container: Container::StructR(Rc::clone(&self.fields)),
-        })))
+    fn fresh(val: ValueImpl) -> PartialVMResult<Self> {
+        match val {
+            ValueImpl::Container(Container::StructR(fields)) => Ok(Self::Fresh { fields }),
+            _ => Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("failed to publish fresh: not a resource".to_string()),
+            ),
+        }
     }
 
-    pub fn mark_dirty(&self) -> PartialVMResult<()> {
-        *self.status.borrow_mut() = GlobalDataStatus::Dirty;
+    fn move_from(&mut self) -> PartialVMResult<ValueImpl> {
+        let fields = match self {
+            Self::None | Self::Deleted => {
+                return Err(PartialVMError::new(StatusCode::MISSING_DATA))
+            }
+            Self::Fresh { .. } => match std::mem::replace(self, Self::None) {
+                Self::Fresh { fields } => fields,
+                _ => unreachable!(),
+            },
+            Self::Cached { .. } => match std::mem::replace(self, Self::Deleted) {
+                Self::Cached { fields, .. } => fields,
+                _ => unreachable!(),
+            },
+        };
+        if Rc::strong_count(&fields) != 1 {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("moving global resource with dangling reference".to_string()),
+            );
+        }
+        Ok(ValueImpl::Container(Container::StructR(fields)))
+    }
+
+    fn move_to(&mut self, val: ValueImpl) -> PartialVMResult<()> {
+        match self {
+            Self::Fresh { .. } | Self::Cached { .. } => {
+                return Err(PartialVMError::new(
+                    StatusCode::CANNOT_WRITE_EXISTING_RESOURCE,
+                ))
+            }
+            Self::None => *self = Self::fresh(val)?,
+            Self::Deleted => *self = Self::cached(val, GlobalDataStatus::Dirty)?,
```

Comment:
@vgao1996: I guess my question is: if you replace this with `Self::fresh { val }` and delete the `status` field altogether, what would go wrong?

Reply:
Ok, I think I see the answer :). Sorry, just ramping up and trying to understand this...

Reply:
`Deleted` will never become `Fresh`. Only the following conversions are allowed through move_to/move_from: `Deleted` <=> `Cached`,  `None <=> Fresh` and mutability through `borrow_global_mut` only allow `Cached { Clean } => Cached { Dirty }`.